feat : add follow/unfollow functionality

achalbajpai opened the pull request:
### What 
Solves #852 

- [x] Allow users to follow/unfollow a conversation
- [x] Send email notification when a new event takes place in the conversation
- [x] Added detailed E2E tests
- [x] Added unit tests 

### Testing 

https://github.com/user-attachments/assets/3b9f8e79-c59e-4e81-a295-d517be7d6653


### Test Cases 
<img width="1184" height="389" alt="Screenshot 2025-08-06 at 6 19 06 PM" src="https://github.com/user-attachments/assets/4ae9f3c7-81fe-4821-ba0f-1e810260f892" />




<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Follow/unfollow conversations via a Follow button in the conversation UI (dynamic state, tooltip, disabled during loading).
  * Followers receive email notifications for new messages, status changes, assignment changes, and notes.

* **Bug Fixes**
  * Improved toast feedback and error handling for follow/unfollow actions.

* **Tests**
  * E2E and unit tests covering follow/unfollow flows, notification sending, persistence, and error cases.

* **Chores**
  * Added follower storage, notification job/event, email template, and related DB/index updates.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

achalbajpai commented on lib/emails/followerNotification.tsx:
```diff
@@ -0,0 +1,174 @@
+      return {
+        title: "Conversation status changed",
+        content: (
+          <Text>
+            <strong>{triggeredByName}</strong> changed the status: <span>{eventDetails.oldStatus}</span> →{" "}
+            <strong>{eventDetails.newStatus}</strong>
+          </Text>
+        ),
+      };
+    case "assignment_change":
+      return {
+        title: "Conversation assignment changed",
+        content: (
+          <Text>
+            <strong>{triggeredByName}</strong> changed the assignment:{" "}
+            <span>{eventDetails.oldAssignee || "Unassigned"}</span> →{" "}
+            <strong>{eventDetails.newAssignee || "Unassigned"}</strong>
+          </Text>
+        ),
+      };
+    case "note_added":
+      return {
+        title: "New note added",
+        content: (
+          <Text>
+            <strong>{triggeredByName}</strong> added a note to this conversation.
+          </Text>
+        ),
+      };
+    default:
+      return {
+        title: "Conversation updated",
+        content: (
+          <Text>
+            <strong>{triggeredByName}</strong> made changes to this conversation.
+          </Text>
+        ),
+      };
+  }
+};
```
> _🛠️ Refactor suggestion_
> 
> **Consider sanitizing message content for security.**
> 
> The message content is directly rendered without HTML sanitization, which could potentially lead to XSS issues if message content contains malicious HTML.
> 
> Consider sanitizing the message content before rendering:
> 
> ```diff
> +import DOMPurify from 'isomorphic-dompurify';
> 
>  {eventDetails.message && (
>    <div
>      style={{
>        background: "#f5f5f5",
>        padding: "10px",
>        borderLeft: "3px solid #007bff",
>        margin: "12px 0",
>        fontSize: "0.875rem",
>      }}
>    >
> -    {eventDetails.message.substring(0, 200)}
> +    {DOMPurify.sanitize(eventDetails.message).substring(0, 200)}
>      {eventDetails.message.length > 200 ? "..." : ""}
>    </div>
>  )}
> ```
> 
> 📝 Committable suggestion
> 🤖 Prompt for AI Agents

This is email content rendered by React which auto-escapes text, and email clients don't execute JavaScript imo , so not needed ig

achalbajpai commented:
Hi @slavingia @binary-koan , Can I get a review on this PR , thank you!

achalbajpai commented:
Hi @slavingia , as suggested I have added the E2E tests for the following implementation.

<img width="864" height="911" alt="image" src="https://github.com/user-attachments/assets/f343bd66-29e1-49a7-9adb-5adf924e6722" />

achalbajpai commented on app/(dashboard)/[category]/conversation/conversationSidebar.tsx:
```diff
@@ -119,6 +120,10 @@ const ConversationSidebar = ({ conversation }: ConversationSidebarProps) => {
           <div className="min-w-0">
             <IssueAssignButton initialIssueGroupId={conversation.issueGroupId} />
           </div>
+          <span className="text-muted-foreground">Follow</span>
```
> Personally I don't think this is in the right spot. Can we move it to the bottom right of the messaging window?
> 
> <img alt="Image" width="513" height="202" src="https://private-user-images.githubusercontent.com/74396/475132690-4bf83998-2987-402b-aef3-ea8c7ec89588.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTQ1MDA3OTcsIm5iZiI6MTc1NDUwMDQ5NywicGF0aCI6Ii83NDM5Ni80NzUxMzI2OTAtNGJmODM5OTgtMjk4Ny00MDJiLWFlZjMtZWE4YzdlYzg5NTg4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODA2VDE3MTQ1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI1NTJjY2VlZGJlZDZiYjk0OGZiYmNhMWQ2NTUzOTk4MjhjMDI3Nzc5ZmEwNzA3ZDFmYTA3MjgxMDliNDg2NjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.XFdcL9YOdfXVaAsP0x5sbY_z4il4OhA2dPpsnkPF9w8">

Sure, I’ll update it here.

binary-koan commented on lib/data/conversationMessage.ts:
```diff
@@ -411,6 +411,21 @@ export const createConversationMessage = async (
+          eventType: "new_message" as const,
+          triggeredByUserId: message.userId,
+          eventDetails: {
+            message: message.body || undefined,
```
Since we're only showing a preview anyway, `cleanedUpText` might be better than `body`

binary-koan commented on lib/data/note.ts:
```diff
@@ -36,6 +37,16 @@ export const addNote = async ({
+        conversationId,
+        eventType: "note_added" as const,
+        triggeredByUserId: user.id,
+        eventDetails: {},
```
Can we include the note body here like with messages?

binary-koan commented on lib/emails/followerNotification.tsx:
```diff
@@ -0,0 +1,174 @@
+    <Html>
+      <Head />
+      <Preview>
+        {title}: {conversationSubject}
```
```suggestion
        {title}
```

We already have the conversation subject in the email subject

binary-koan commented on tests/e2e/conversations/follow-conversation.spec.ts:
```diff
@@ -0,0 +1,313 @@
+
+
+  test("should display follow button in conversation sidebar", async ({ page }) => {
+    const conversationsPage = new ConversationsPage(page);
```
We're currently removing all the page objects including this one; could you check?

achalbajpai replied:
Sure will switch to Playwright integration

binary-koan commented on tests/e2e/conversations/follower-notifications.spec.ts:
```diff
@@ -0,0 +1,295 @@
+  });
+
+
+  test("should trigger follower notification on new message", async ({ page }) => {
```
We're not actually asserting the notification is sent here, are we? If not then I don't think these tests are necessary, the other file is enough

binary-koan reviewed (commented):
Thanks for working on this! Added some comments.

Could you also please
- Screenshot the email + mobile conversation view
- Add unit tests for the router and job

achalbajpai commented:
### Desktop View 
<img width="1485" height="799" alt="Screenshot 2025-08-09 at 4 12 07 PM" src="https://github.com/user-attachments/assets/ff0719d2-afbc-4408-b63f-21c2e3126dc5" />

### Mobile View 
<img width="355" height="727" alt="Screenshot 2025-08-09 at 4 06 30 PM" src="https://github.com/user-attachments/assets/e224a034-60a7-4fbb-b721-a858393191e8" />

achalbajpai commented:
> Thanks for working on this! Added some comments.
> 
> Could you also please
> 
> * Screenshot the email + mobile conversation view
> * Add unit tests for the router and job

Hi @binary-koan , thanks for the detailed review , I have addressed all the comments 
- Added unit tests
- Added screenshot as well for mail view

achalbajpai commented on app/(dashboard)/[category]/conversation/followButton.tsx:
```diff
@@ -0,0 +1,95 @@
+      <Tooltip>
+        <TooltipTrigger asChild>
+          <Button
+            variant={isFollowing ? "default" : "outlined"}
```
Outlined is correct , our codebase has it

> _⚠️ Potential issue_
> 
> **Fix Button variant: use "outline" (not "outlined")**
> 
> shadcn/ui uses "outline". "outlined" will not match the variant union and may break styles/types.
> 
> ```diff
> -            variant={isFollowing ? "default" : "outlined"}
> +            variant={isFollowing ? "default" : "outline"}
> ```
> 
> 📝 Committable suggestion
> 🤖 Prompt for AI Agents
> ```
> In app/(dashboard)/[category]/conversation/followButton.tsx at line 75, the
> Button component's variant prop is incorrectly set to "outlined". Change the
> variant value from "outlined" to "outline" to match the accepted variant options
> in shadcn/ui and ensure proper styling and type safety.
> ```

achalbajpai commented on jobs/sendFollowerNotification.ts:
```diff
@@ -0,0 +1,142 @@
+      },
+    });
+
+    if (followers.length === 0) {
+      return;
+    }
```
No changes needed - the function correctly returns undefined for no-op cases which is well-tested , what do you think

binary-koan reviewed (commented):
Thanks! Added some more minor comments, looks good apart from that

achalbajpai commented on jobs/sendFollowerNotification.ts:
```diff
@@ -0,0 +1,142 @@
+
+    const triggeredByName = triggeredByUser?.displayName || triggeredByUser?.user?.email || "Someone";
+
+    const conversationLink = `${env.AUTH_URL}/conversations/${conversation.slug}`;
```
Agreed, that would be much cleaner

achalbajpai commented on lib/data/note.ts:
```diff
@@ -36,6 +37,18 @@ export const addNote = async ({
+        eventType: "note_added" as const,
+        triggeredByUserId: user.id,
+        eventDetails: {
+          note: note.body || undefined,
```
Not used now

binary-koan replied:
My original request was to show this in the email; pushed a small fix to restore this and show it

achalbajpai replied:
> My original request was to show this in the email; pushed a small fix to restore this and show it

Some confusion from my end , thanks for fixing it up :)

achalbajpai commented:
> Thanks! Added some more minor comments, looks good apart from that

Addressed all the comments , hope the PR is good to go now :) Let me know if any other changes are needed.

achalbajpai commented:
> Thanks! Added some more minor comments, looks good apart from that

Addressed all the comments @binary-koan  , hope the PR is good to go now :) Let me know if any other changes are needed.

slavingia commented:
Adding priority label, if this isn't good enough to merge tmmrw probably best to do internally.
